stdlib/src/memory/arc.mojo

```diff
@@ -45,32 +45,231 @@ print(Arc(String("ok"))[])
 """
 
 from os.atomic import Atomic
+from os import abort
 
 from builtin.builtin_list import _lit_mut_cast
 from memory import UnsafePointer, stack_allocation
+from memory.maybe_uninitialized import UnsafeMaybeUninitialized
+from collections import Optional, OptionalReg
 
 
-struct _ArcInner[T: Movable]:
+struct _ArcInner[T: Movable, enable_weak: Bool = False]:
     var refcount: Atomic[DType.uint64]
-    var payload: T
+    var payload: UnsafeMaybeUninitialized[T]
+    
+    alias _weak_refcount_type = __mlir_type[
+        `!pop.array<`, Int(enable_weak).value, `, `, Atomic[DType.int64], `>`
+    ]
```

Comment:
**Comment** This is clever using the boolean to make this a zero-sized field, neat trick! 🙂